test/python/circuit/test_instruction_power.py

```diff
@@ -0,0 +1,225 @@
+
+        expected = QuantumCircuit(qr)
+        expected.append(U3Gate(0, 0, pi / 2), qr[:])
+
+        self.assertEqual(result, expected)
+
+    def test_unroller_zero(self):
+        """Test unrolling gate.power(0).
+        """
+        qr = QuantumRegister(1, 'qr')
+
+        circuit = QuantumCircuit(qr)
+        circuit.append(SGate().power(0), qr[:])
+        result = PassManager(Unroller('u3')).run(circuit)
+
+        expected = QuantumCircuit(qr)
+        expected.append(U3Gate(0, 0, 0), qr[:])
+
+        self.assertEqual(result, expected)
+
+    def test_unroller_minus_one(self):
+        """Test unrolling gate.power(-1).
+        """
+        qr = QuantumRegister(1, 'qr')
+
+        circuit = QuantumCircuit(qr)
+        circuit.append(SGate().power(-1), qr[:])
+        result = PassManager(Unroller('u3')).run(circuit)
+
+        expected = QuantumCircuit(qr)
+        expected.append(U3Gate(0, 0, -pi / 2), qr[:])
+
+        self.assertEqual(result, expected)
+
+    def test_unroller_minus_two(self):
+        """Test unrolling gate.power(-2).
+        """
+        qr = QuantumRegister(1, 'qr')
+
+        circuit = QuantumCircuit(qr)
+        circuit.append(SGate().power(-2), qr[:])
+        result = PassManager(Unroller('u3')).run(circuit)
+
+        expected = QuantumCircuit(qr)
+        expected.append(U3Gate(0, 0, -pi / 2), qr[:])
+        expected.append(U3Gate(0, 0, -pi / 2), qr[:])
+
+        self.assertEqual(result, expected)
+
+
+class TestGateSqrt(QiskitTestCase):
+    """Test square root using Gate.power()"""
+
+    def test_unitary_sqrt(self):
+        """Test UnitaryGate.power(1/2) method.
+        """
+        expected = array([[0.5 + 0.5j, 0.5 + 0.5j],
+                          [-0.5 - 0.5j, 0.5 + 0.5j]])
+        result = UnitaryGate([[0, 1j], [-1j, 0]]).power(1 / 2)
+        self.assertEqual(result.name, 'unitary^0.5')
+        self.assertEqual(len(result.definition), 1)
+        assert_allclose(result.definition[0][0].to_matrix(), expected)
+
+    def test_starndard_sqrt(self):
+        """Test standard Gate.power(1/2) method.
+        """
+        expected = array([[1 + 0.j, 0 + 0.j],
+                          [0 + 0.j, 0.70710678 + 0.70710678j]])
+        result = SGate().power(1 / 2)
+        self.assertEqual(result.name, 's^0.5')
+        self.assertEqual(len(result.definition), 1)
+        assert_allclose(result.definition[0][0].to_matrix(), expected)
+
+
+@ddt
+class TestGateFloat(QiskitTestCase):
+    """Test power generalization to root calculation"""
+
+    @data(2, 3, 4, 5, 6, 7, 8, 9)
+    def test_direct_root(self, degree):
+        """Test nth root"""
+        result = SGate().power(1 / degree)
+        self.assertEqual(result.name, 's^'+str(1/degree))
+        self.assertEqual(len(result.definition), 1)
+        assert_allclose(matrix_power(result.definition[0][0].to_matrix(), degree),
+                        SGate().to_matrix())
+
+    @data(2.1, 3.2, 4.3, 5.4, 6.5, 7.6, 8.7, 9.8)
```

Comment:
Do we have a test for floats < 1, negative?

Reply:
Test added in d26a05ca.

It seems that `-0.2` is not working @eddieschoute 

```
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

x and y nan location mismatch:
 x: array([[1.      +0.j      ,      nan     +nanj],
       [     nan     +nanj, 0.951057-0.309017j]])
 y: array([[1.      +0.j      , 0.      +0.j      ],
       [0.      +0.j      , 0.951057-0.309017j]])
```

Reply:
There are NaNs in there so something was divided by zero probably

Reply:
The `NaN`s are from `SGate().to_matrix() ** exponent`

Reply:
Fixed in e6069b4e